File: Ladder02/Python_code/recover_rotated_sorted_array.py
class Solution:
    """
    @param nums: The rotated sorted array
    @return: nothing
    """
    def recoverRotatedSortedArray(self, nums):
        if not nums or nums[0] < nums[-1]:
            return
        start = 0
        end = len(nums) - 1
        while start + 1 < end:
            mid = start + (end - start) / 2
            if nums[mid] > nums[end]:
                start = mid
            elif nums[mid] < nums[start]:
                end = mid
            else:
                end -= 1
        pivot = start if nums[start] < nums[end] else end
        self.reverse(nums, pivot, len(nums) - 1)
        self.reverse(nums, 0, pivot - 1)
        self.reverse(nums, 0, len(nums) - 1)

    def reverse(self, A, start, end):
        while start < end:
            A[start], A[end] = A[end], A[start]
            start, end = start + 1, end - 1

def main():
    test = Solution()
    nums = [1, 1, 1, 1, 1, 1, 1, 1, 1, -1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]
    test.recoverRotatedSortedArray(nums)

if __name__ == '__main__':
    main()

File: NineChapters/02.BinarySearch_SortedArray/Java_Code/RecoverRotatedSortedArray.java
import java.util.ArrayList;
import java.util.Arrays;

public class RecoverRotatedSortedArray {
    /**
     * @param nums: The rotated sorted array
     * @return: void
     */
    public void recoverRotatedSortedArray(ArrayList<Integer> nums) {
        if (nums == null || nums.size() == 0
            || nums.get(0) < nums.get(nums.size() - 1)) {
            return;
        }
        int start = 0;
        int end = nums.size() - 1;
        int mid;
        while (start + 1 < end) {
            mid = start + (end - start) / 2;
            if (nums.get(mid) < nums.get(end)) {
                end = mid;
            }
            else if (nums.get(mid) > nums.get(end)) {
                start = mid;    
            }
            else {
                end--;
            }
        }
        int pivot = (nums.get(start) < nums.get(end)) ? start : end;
        reverse(nums, 0, pivot - 1);
        reverse(nums, pivot, nums.size() - 1);
        reverse(nums,0, nums.size() - 1);
    }

    private void reverse(ArrayList<Integer> nums, int start, int end) {
        while (start < end) {
            int temp = nums.get(start);
            nums.set(start, nums.get(end));
            nums.set(end, temp);
            start++;
            end--;
        }
    }

    public static void main(String[] args) {
        RecoverRotatedSortedArray test = new RecoverRotatedSortedArray();
        ArrayList<Integer> nums = new ArrayList<Integer>(Arrays.asList(
            1, 1, 1, 1, 1, 1, 1, 1, 1, -1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1));
        test.recoverRotatedSortedArray(nums);
    }
}